### PR TITLE
First attempt Unit Testing

### PR DESCRIPTION
**Pull request by zhenhuaizeng**
We did Unit Test on methods (info, debug, annoy, warn, error, fatal, and get_logging_path).

### CODE REVIEW

**Comment by jrolstad** [comment]
Recommend adding the description of the Pull Request in comment above to the PR description so it's easier to find.  Also, the description needs to give a high level summary of what changed (which you did), but also why

Example:
_Adding unit tests around the existing slogpy public logging methods focused on validating output to the log file and console.  Tests are intended to support upcoming changes to prevent regressions._

File: tests/mytest.py
Comment (jrolstad) [inline tests/mytest.py:1]: Recommend renaming this file to be more descriptive

**Comment by victorcfk** [comment]
Hey @zhenhuaizeng I think we discussed having the tests also sharing the same input. In the case of the tests above, all of them are using a different input, which means that you are potentially testing string combinations as well. This matters because things like escape characters or symbols (e.g. "!") can behave differently when used in strings. I believe Alex from the team brought this up.

I would suggest you test that separately as well in a different pull request.

Additionally, some documentation, what was the task/bug that you are working on that is related to the pull request? (e.g link to the basecamp task, or the documentation that Sagal is creating)

File: tests/__init__.py
Comment (neel-m) [inline tests/__init__.py:1]: What is the purpose of the content in this file?
Dunder init.py is what makes a folder a Python package (pre Python 3.3, but still frequently used).  The only content that should be here is content that is needed for initialization.

File: tests/mytest.py
Comment (neel-m) [reply tests/mytest.py:1]: More importantly, what is the purpose of this file here?
Pytest will not do anything with it and this isn't a unit test.

File: tests/test_slogpy.py
Comment (neel-m) [inline tests/test_slogpy.py:1]: Since this file appears to only test console output, it might make sense to call it something like test_console.py

File: tests/test_slogpy.py [hunk 1]
```diff
@@ -1,8 +1,73 @@
...
+
+def test_logging_info(capsys):
+
+    slog.info("Log something at the info level")
+    captured = capsys.readouterr()
+     # Remove ANSI escape codes using regex
+    cleaned_output = re.sub(r'\x1b\[[0-9;]*m', '', captured.out)
+    assert cleaned_output == "Log something at the info level\n"
+
+def test_logging_debug(capsys):
```
Comment (neel-m) [inline tests/test_slogpy.py]: This doesn't really test what is important about debug (that debug output is NOT sent to the console under normal circumstances)

File: tests/test_slogpy.py [hunk 2]
```diff
@@ -1,8 +1,73 @@
...
+
+    slog.debug('log some debugging...this will only go to file unless you set the logging level')
+    captured = capsys.readouterr()
+    cleaned_output = re.sub(r'\x1b\[[0-9;]*m', '', captured.out)
+    assert cleaned_output == "DEBUG: log some debugging...this will only go to file unless you set the logging level\n"
+
+    # Reset logging level (optional)
+    slog.set_log_level(original_level)
+
+def test_logging_annoy(capsys):
```
Comment (neel-m) [inline tests/test_slogpy.py]: this doesn't really test what is special about annoy (you get it only once)

File: tests/test_slogpy.py [hunk 3]
```diff
@@ -1,8 +1,73 @@
...
 def test_version():
     """The most basic of tests."""
     assert __version__ == "0.1.0"
+
+def test_logging_info(capsys):
+
+    slog.info("Log something at the info level")
+    captured = capsys.readouterr()
+     # Remove ANSI escape codes using regex
+    cleaned_output = re.sub(r'\x1b\[[0-9;]*m', '', captured.out)
```
Comment (neel-m) [inline tests/test_slogpy.py]: For most people regex's are not readable, even with regex101 this one is not understandable.  This one is likely to strip the content that rich adds to get colors.  It is also duplicated (cut-n-paste) many times.  This should be set as a file level constant that has a meaningful name.

**Review by neel-m — COMMENTED** [review]
the new code is not ruff check clean

File: tests/mytest.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        for tick in progress.track(range(0,10093), description='pick fruit'):
+            if tick % 427 == 0:
+                slog.info(f'{tick} mod 7 is 0')
+            time.sleep(0.0003)
+
+
+if __name__ == '__main__':
+    test_progress()
+
+def some_function(name, title, number=27):
```
Comment (jrolstad) [inline tests/mytest.py]: This isn't a test, should probably be removed.  These files should have unit tests only, methods such as these to show sample usage are helpful but belong elsewhere in an examples directory or project wiki

File: tests/mytest.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+# if you want to tell the user where the log is (after any logging!)
+print(f'Log written to: {slog.get_logging_path()}')
+
+# You can call SlogProgress() just like you would rich.progress.Progress
+# or you can use one of the predefined factory methods. If you have a
+# progress bar that you use often, add a factory for it in
+# slogpy/slogpy/progress.py
+
+def test_progress():
```
Comment (jrolstad) [inline tests/mytest.py]: This is a test without assertions.  Recommend verifying (asserting) behavior or removing

File: tests/mytest.py
Comment (jrolstad) [reply tests/mytest.py:1]: @zhenhuaizeng , please add the summary to the PR description so we can resolve this comment

Comment (jrolstad) [reply tests/mytest.py:1]: @zhenhuaizeng , if this file isn't needed, please remove

File: tests/test_slogpy_console.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
+"""The test for the slogpy package."""
+from slogpy import slog
+from slogpy import __version__
+from slogpy.slog import Slog as slog
+def test_version():
```
Comment (jrolstad) [inline tests/test_slogpy_console.py]: This should probably be in a separate file since it's not testing console output

**Comment by jrolstad** [comment]
@zhenhuaizeng see my comment above about filling out the description of the PR https://github.com/slogpy/slogpy/pull/3#issuecomment-2081531205

**Comment by zhenhuaizeng** [comment]
Adding unit tests around the existing slogpy public logging methods focused on validating output to the log file and console. Giving 'happy path' for each of the methods on the public interface(info,debug,annoy,warn,error,fatal) to ensure they log the expected message.

**Comment by zhenhuaizeng** [comment]
We removed the "test_version" method from the test_slogpy_console.py file because it was not testing console output.

File: tests/__init__.py
Comment (victorcfk) [inline tests/__init__.py:1]: Hey @zhenhuaizeng for this commit, discard the changes here, its not necessary and will mislead reviewers and future code comitters

File: tests/test_slogpy_file.py
Comment (victorcfk) [inline tests/test_slogpy_file.py:1]: Hi @zhenhuaizeng i think you should keep this commit in a separate pull request. The goal of the first assignment was just about testing the console output. I think we can make it so that the entire pull request is just about your first task, unit test the **console**, not the **console and the file**

Comment (zhenhuaizeng) [reply tests/test_slogpy_file.py:1]: Can you give me some instructions on how to do it? I assume that I should delete the test_slogpy_file in my VS code, make the pull request, restore test_slogpy_file, and then make a pull request again?

Comment (victorcfk) [reply tests/test_slogpy_file.py:1]: The simplest is to amend this existing commit(and save your file unit test changes somewhere else), so that your commit just contains the change for console log, then push the changed commit. 

You can also try splitting the commit in command line or your visual git editor and push the changed commit.

Comment (victorcfk) [reply tests/test_slogpy_file.py:1]: Yes, delete test_slogpy_file.py, make a commit and amend this current pull request. 

Restore the test_slogpy_file.py and make a commit with that file in a separate pull request.

**Review by victorcfk — COMMENTED** [review]
Please use the "changes from all commits" view to understand the changes you are making. Codebase owners / reviewers are likely to want to see that your entire pull request makes sense before looking at each individual commit.

**Comment by zhenhuaizeng** [comment]
Removed test_slogpy_file.py from the VS code and uploaded test_slogpy_console to Git Hub because our first task was to focus on the console's unit tests.

**Comment by zhenhuaizeng** [comment]
Restored the test_slogpy_file to the VS code and uploaded the test_slogpy_file to Git Hub as our second task, as described on Basecamp.

File: tests/__init__.py
Comment (victorcfk) [inline tests/__init__.py:1]: @zhenhuaizeng seems like there was a whitespace removal here, can you revert that?

File: tests/test_slogpy.py
Comment (victorcfk) [reply tests/test_slogpy.py:1]: @zhenhuaizeng same comment as above, whitespace removal

File: tests/__init__.py
Comment (zhenhuaizeng) [reply tests/__init__.py:1]: what whitespace? is it in __init__.py file?

File: tests/test_slogpy_console.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
+"""Import slogpy from slog"""
+from slogpy.slog import Slog as slog # noqa: N813 I001
```
Comment (neel-m) [inline tests/test_slogpy_console.py]: This project enforces style with ruff.  Generally speaking it is not okay to bypass the rules.  Please remove I001 and fix the code to match the style, ruff can probably do that for you automatically.

File: tests/test_slogpy_console.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+"""This is a setup function"""
+def setup_function():
+    """Not to create a file for output"""
```
Comment (neel-m) [inline tests/test_slogpy_console.py]: This sentence is not very clear. I think I understand what it means.

File: tests/test_slogpy_console.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    slog.initialize(file_logging=False)
+
+"""First unit test"""
```
Comment (neel-m) [inline tests/test_slogpy_console.py:9]: These comments are unnecessary and the use of triple quote is not appropriate for a single line comment.  But, https://peps.python.org/pep-0257/ says that using triple double quotes for a single line in a docstring is common practice.

File: tests/test_slogpy_console.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+def test_info_console_output(capsys, msg="This is a console test"):
+    """Passing a message to the info method and then capture the message from the console"""
+    slog.info(msg)
+    expected_content = f"{msg}\n"
+    out, _ = capsys.readouterr()
+    assert expected_content == out
+
+"""Second unit test (Debug Method)"""
+def test_debug_console_output(capsys, msg="This is a console test"):
+    """Set the log level to 1, so the output can be captured from the consol"""
```
Comment (neel-m) [inline tests/test_slogpy_console.py]: The better test for debug (for console) is that there is no output to the console.  This is the special thing about debug, the content only goes to the log and not to the console.

The thing this test is testing is that it is possible to set the level to something else and that it works.

File: tests/test_slogpy_console.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    slog.set_log_level(1)
```
Comment (neel-m) [inline tests/test_slogpy_console.py]: Slog (the class) provides handy constants to make it clearer what level is being set:
` slog.set_log_level(slog.DEBUG)`

File: tests/test_slogpy_file.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
+"""Import slogpy from slog"""
+from slogpy.slog import Slog as slog # noqa: N813 I001
```
Comment (neel-m) [inline tests/test_slogpy_file.py]: same comment about I001

File: tests/test_slogpy_file.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+import tempfile
+
+"""This is a setup function"""
+def setup_function():
+    """create a file for output"""
```
Comment (neel-m) [inline tests/test_slogpy_file.py]: Why is this here, when what it does is overwritten by the initialize in the next function?

File: tests/test_slogpy_file.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    slog.initialize(file_logging=True)
+
+"""debug method unit test"""
+def test_debug_file(msg="This is a file test"):
+    """The test for the debug function."""
```
Comment (neel-m) [inline tests/test_slogpy_file.py:8]: Since debug is special, it might make sense to test it specifically.  Since this test is about logging to file, perhaps it would make sense to start with info.

**Review by neel-m — CHANGES_REQUESTED** [review]
Getting closer and just a few changes that need to be made.

File: tests/test_slogpy_file.py (hunk 13 above)
Comment (zhenhuaizeng) [reply tests/test_slogpy_file.py:7]: I don't know why if I don't initialize file_logging = True (slog.initialize(file_logging=True)), it will not create a log file when I run pytest.

File: tests/test_slogpy_file.py (hunk 14 above)
Comment (zhenhuaizeng) [reply tests/test_slogpy_file.py:8]: I'm not sure if I understood this comment correctly because Alex was delegated to do a unit test for slog.info. What did you mean "start with info" and "test it specifically". I need more clarification. Thank you!

File: tests/test_slogpy_console.py (hunk 7 above)
Comment (zhenhuaizeng) [reply tests/test_slogpy_console.py]: ruff didn't do that for me. Instead, I installed and used "isort" to help me fix it.

File: tests/test_slogpy_file.py (hunk 13 above)
Comment (zhenhuaizeng) [reply tests/test_slogpy_file.py]: My bad. I found out why...

**Comment by neel-m** [comment]
Closing this as it is now very old and obsolete.
